hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java

```diff
@@ -0,0 +1,390 @@
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws Exception {
+        NodeEngine nodeEngine = ((HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance()).node.nodeEngine;
+
+        traverser = new IMapIndexTraverser(
+                nodeEngine,
+                indexScanMetadata,
+                this.partitions,
+                SimpleExpressionEvalContext.from(context)
+        );
+    }
+
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(traverser);
+    }
+
+    static class IMapIndexTraverser implements Traverser<Object[]> {
+        private final MapContainer mapContainer;
+        private final List<Expression<?>> projections;
+        private final Expression<Boolean> filter;
+        private PartitionIdSet partitions;
+        private final MapScanRow row;
+        private final SimpleExpressionEvalContext ctx;
+
+        private InternalIndex index;
+
+        private final String indexName;
+        private final IndexFilter indexFilter;
+        private final List<QueryDataType> converterTypes;
+        private final int componentCount;
+        // The collation of index entries
+        private List<Boolean> ascs;
+
+        /**
+         * Stamp to ensure that indexed partitions are stable throughout query execution.
+         */
+        private Long partitionStamp;
+
+        private final KeyValueIterator iteratorExec;
+
+        IMapIndexTraverser(
+                @Nonnull final NodeEngine nodeEngine,
+                @Nonnull final MapIndexScanMetadata indexScanMetadata,
+                @Nonnull final PartitionIdSet partitions,
+                @Nonnull SimpleExpressionEvalContext ctx
+
+        ) {
+            this.ctx = ctx;
+            MapService mapService = nodeEngine.getService(MapService.SERVICE_NAME);
+            this.mapContainer = mapService.getMapServiceContext().getMapContainer(indexScanMetadata.getMapName());
+            this.projections = indexScanMetadata.getProjects();
+            this.filter = indexScanMetadata.getFilter();
+
+            this.indexName = indexScanMetadata.getIndexName();
+            this.indexFilter = indexScanMetadata.getIndexFilter();
+            this.converterTypes = indexScanMetadata.getConverterTypes();
+            this.componentCount = indexScanMetadata.getIndexComponentCount();
+            this.ascs = indexScanMetadata.getAscs();
+
+            this.iteratorExec = createIterator();
+            this.row = MapScanRow.create(
+                    indexScanMetadata.getKeyDescriptor(),
+                    indexScanMetadata.getValueDescriptor(),
+                    indexScanMetadata.getFieldPaths(),
+                    indexScanMetadata.getFieldTypes(),
+                    mapContainer.getExtractors(),
+                    ctx.getSerializationService()
+            );
+        }
+
+        @Override
+        public Object[] next() {
+            while (iteratorExec.tryAdvance()) {
```

Comment:
We should  limit the number of rows scanned in one call to not break the cooperative contract, similarly to `OnHeapMapScanP`'s traverser.